File: supabase/enrich-places.mjs
// Fetches reviews, rating, and photos for each place from the Google Places
// API (New), asks Claude to pull out commonly-mentioned dishes from those
// reviews, and writes it all into the `places` table for detail.html to
// read. Never called from the browser — the API keys below must stay
// server-side (Yelp explicitly forbids client-side calls, and Google's key
// would be exposed too).
//
// Usage:
//   cd supabase
//   npm install
//   cp .env.example .env   # fill in all four vars
//   npm run enrich
//
// Safe to re-run: only places with enriched_at still null (never
// successfully enriched) are processed. Delete enriched_at on a row (or run
// the SQL: update places set enriched_at = null) to force a refresh.

import { createClient } from "@supabase/supabase-js";
import "dotenv/config";

const supabaseUrl = process.env.SUPABASE_URL;
const serviceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
const googleApiKey = process.env.GOOGLE_PLACES_API_KEY;
const anthropicApiKey = process.env.ANTHROPIC_API_KEY;

if (!supabaseUrl || !serviceRoleKey || !googleApiKey || !anthropicApiKey) {
  console.error(
    "Missing one of SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, GOOGLE_PLACES_API_KEY, ANTHROPIC_API_KEY. Copy .env.example to .env and fill it in."
  );
  process.exit(1);
}

const supabase = createClient(supabaseUrl, serviceRoleKey);

const PLACE_PHOTO_FIELD_MASK =
  "places.id,places.rating,places.userRatingCount,places.reviews,places.photos,places.websiteUri,places.googleMapsUri";

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function searchGooglePlace(place) {
  const textQuery = [place.name, place.neighborhood, "San Francisco, CA"].filter(Boolean).join(", ");

  const res = await fetch("https://places.googleapis.com/v1/places:searchText", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Goog-Api-Key": googleApiKey,
      "X-Goog-FieldMask": PLACE_PHOTO_FIELD_MASK,
    },
    body: JSON.stringify({ textQuery, maxResultCount: 1 }),
  });

  if (!res.ok) {
    throw new Error(`Places API search failed (${res.status}): ${await res.text()}`);
  }

  const data = await res.json();
  return data.places?.[0] ?? null;
}

async function uploadPhotos(placeId, photos) {
  const urls = [];

  for (const [i, photo] of (photos ?? []).slice(0, 6).entries()) {
    // Ask for the CDN URL as JSON instead of a 302 redirect (skipHttpRedirect)
    // and fetch that URL directly — some sandboxed environments don't follow
    // the redirect from the media endpoint.
    const lookupRes = await fetch(
      `https://places.googleapis.com/v1/${photo.name}/media?maxWidthPx=1200&key=${googleApiKey}&skipHttpRedirect=true`
    );
    if (!lookupRes.ok) {
      console.warn(`  photo ${i} lookup failed (${lookupRes.status}), skipping`);
      continue;
    }
    const { photoUri } = await lookupRes.json();

    const mediaRes = await fetch(photoUri);
    if (!mediaRes.ok) {
      console.warn(`  photo ${i} fetch failed (${mediaRes.status}), skipping`);
      continue;
    }

    const contentType = mediaRes.headers.get("content-type") ?? "image/jpeg";
    const bytes = new Uint8Array(await mediaRes.arrayBuffer());
    const path = `${placeId}/${i}.jpg`;

    const { error: uploadError } = await supabase.storage
      .from("place-photos")
      .upload(path, bytes, { contentType, upsert: true });
    if (uploadError) {
      console.warn(`  photo ${i} upload failed: ${uploadError.message}`);
      continue;
    }

    const { data } = supabase.storage.from("place-photos").getPublicUrl(path);
    urls.push(data.publicUrl);
  }

  return urls;
}

function mapReviews(reviews) {
  return (reviews ?? []).slice(0, 5).map((r) => ({
    author: r.authorAttribution?.displayName ?? "Anonymous",
    rating: r.rating ?? null,
    text: r.text?.text ?? r.originalText?.text ?? "",
    relative_time: r.relativePublishTimeDescription ?? "",
  }));
}

async function extractTopDishes(placeName, reviews) {
  if (reviews.length === 0) return [];

  const reviewText = reviews.map((r, i) => `Review ${i + 1} (${r.rating ?? "?"}★): ${r.text}`).join("\n\n");
  const prompt = `Here are Google reviews for "${placeName}", a San Francisco restaurant/cafe:\n\n${reviewText}\n\nBased only on these reviews, list up to 5 specific dishes or menu items that are mentioned positively. Respond with ONLY a JSON array of short strings (e.g. ["Pad See Ew", "Thai Iced Tea"]) and nothing else. If no specific dishes are mentioned, respond with [].`;

  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-api-key": anthropicApiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: "claude-sonnet-5",
      max_tokens: 200,
      messages: [{ role: "user", content: prompt }],
    }),
  });

  if (!res.ok) {
    console.warn(`  top-dish extraction failed (${res.status}): ${await res.text()}`);
    return [];
  }

  const data = await res.json();
  const raw = (data.content?.[0]?.text ?? "[]").replace(/^```(?:json)?\s*|\s*```$/g, "");
  try {
    const dishes = JSON.parse(raw);
    return Array.isArray(dishes) ? dishes.filter((d) => typeof d === "string") : [];
  } catch {
    console.warn(`  couldn't parse top-dish response: ${raw}`);
    return [];
  }
}

const { data: places, error: fetchError } = await supabase
  .from("places")
  .select("id, name, neighborhood")
  .is("enriched_at", null);

if (fetchError) {
  console.error("Failed to load places:", fetchError.message);
  process.exit(1);
}

if (!places || places.length === 0) {
  console.log("Nothing to enrich — every place already has enriched_at set.");
  process.exit(0);
}

console.log(`Enriching ${places.length} place(s)...`);

for (const place of places) {
  console.log(`- ${place.name}`);
  try {
    const match = await searchGooglePlace(place);
    if (!match) {
      console.warn("  no Google Places match found, marking as enriched with no data");
      await supabase.from("places").update({ enriched_at: new Date().toISOString() }).eq("id", place.id);
      continue;
    }

    const reviews = mapReviews(match.reviews);
    const [photos, topDishes] = await Promise.all([
      uploadPhotos(place.id, match.photos),
      extractTopDishes(place.name, reviews),
    ]);

    const { error: updateError } = await supabase
      .from("places")
      .update({
        google_place_id: match.id ?? null,
        google_maps_uri: match.googleMapsUri ?? null,
        rating: match.rating ?? null,
        user_rating_count: match.userRatingCount ?? null,
        reviews,
        photos,
        top_dishes: topDishes,
        enriched_at: new Date().toISOString(),
      })
      .eq("id", place.id);

    if (updateError) {
      console.error(`  failed to save: ${updateError.message}`);
    } else {
      console.log(`  saved (${photos.length} photos, ${reviews.length} reviews, ${topDishes.length} dishes)`);
    }
  } catch (err) {
    console.error(`  error: ${err.message}`);
  }

  await sleep(300);
}

console.log("Done.");

File: supabase/enrich-places.py
import os
from supabase import create_client, Client
import requests
import re
import json
import sys
from datetime import datetime, timezone
from dotenv import load_dotenv

load_dotenv()

# Set up Supabase client using environment variables for URL and API key
url: str = os.environ.get("SUPABASE_URL")
key: str = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
googleApiKey: str = os.environ.get("GOOGLE_PLACES_API_KEY")
anthropicApiKey: str = os.environ.get("ANTHROPIC_API_KEY")


# Create Supabase client and handle potential exceptions
try:
    supabase: Client = create_client(url, key)
except Exception as e:
    raise Exception(f"Failed to create Supabase client: {e}")

PLACE_FIELD_MASK = (
    "places.id,places.rating,places.userRatingCount,places.reviews,places.photos,"
    "places.websiteUri,places.googleMapsUri,places.primaryType,"
    "places.primaryTypeDisplayName,places.types,places.editorialSummary"
)

# Google Places (New) type strings that actually name a cuisine, mapped to a
# display label. Generic types (`restaurant`, `cafe`, `bar`, `bakery`,
# `fine_dining_restaurant`, ...) are deliberately absent — they're a format,
# not a cuisine, so those fall through to Claude.
GOOGLE_CUISINE_LABELS = {
    "afghani_restaurant": "Afghan", "african_restaurant": "African",
    "american_restaurant": "American", "asian_restaurant": "Asian",
    "barbecue_restaurant": "Barbecue", "brazilian_restaurant": "Brazilian",
    "chinese_restaurant": "Chinese", "french_restaurant": "French",
    "greek_restaurant": "Greek", "hamburger_restaurant": "Burgers",
    "indian_restaurant": "Indian", "indonesian_restaurant": "Indonesian",
    "italian_restaurant": "Italian", "japanese_restaurant": "Japanese",
    "korean_restaurant": "Korean", "lebanese_restaurant": "Lebanese",
    "mediterranean_restaurant": "Mediterranean", "mexican_restaurant": "Mexican",
    "middle_eastern_restaurant": "Middle Eastern", "pizza_restaurant": "Pizza",
    "ramen_restaurant": "Ramen", "seafood_restaurant": "Seafood",
    "spanish_restaurant": "Spanish", "steak_house": "Steakhouse",
    "sushi_restaurant": "Sushi", "thai_restaurant": "Thai",
    "turkish_restaurant": "Turkish", "vegan_restaurant": "Vegan",
    "vegetarian_restaurant": "Vegetarian", "vietnamese_restaurant": "Vietnamese",
}


def cuisine_from_google(match: dict):
    """Return a cuisine label from Google's structured place type, or None."""
    if match.get("primaryType") in GOOGLE_CUISINE_LABELS:
        return GOOGLE_CUISINE_LABELS[match["primaryType"]]
    for t in match.get("types", []):
        if t in GOOGLE_CUISINE_LABELS:
            return GOOGLE_CUISINE_LABELS[t]
    return None


def search_google_places(place: dict):
    """
    Search for a place using the Google Places API.

    Args:
        place (dict): A dictionary containing the name and neighborhood

    Returns:
        dict | None: The first matching Google Place, or None
        if no place is found.
    """
    text_query = ", ".join(
        filter(None, [place.get("name"), place.get("neighborhood"), "San Francisco, CA"])
    )


    # Define the API endpoint
    url = 'https://places.googleapis.com/v1/places:searchText'
    # Define the headers
    headers = {
        'Content-Type': 'application/json',
        'X-Goog-Api-Key': googleApiKey,  # Replace 'API_KEY' with your actual Google Places API key
        'X-Goog-FieldMask': PLACE_FIELD_MASK
    }
    
    # Define the data payload for the POST request
    data = {
        "textQuery": text_query,
        "maxResultCount": 1
    }
    # Execute the HTTP POST request
    response = requests.post(url, json=data, headers=headers, timeout = 10)
    # Check if the request was successful
    if not response.ok:
       raise Exception(f"Place API search failed ({response.status_code}): {response.text}")

    places = response.json().get("places", []) 
    return places[0] if places else None  # Return the JSON response from the Google Places API
   
def upload_photos(placeId: str, photos: list):
    """
    Call Google Place API to downloand up to 6 photos of a place, 
    upload photos to Supabase storage, and return public URLs for the uploaded photos.

    Args:
        place_id (str): The ID of the place.
        photos (list): A list of photo dictionaries containing 'photoReference' and 'description'.
    """

    urls = []
    photos = photos or []
    # fetch up to 6 photos
    for i, photo in enumerate(photos[:6]):
        lookupRes = requests.get(f"https://places.googleapis.com/v1/{photo['name']}/media?maxWidthPx=1200&key={googleApiKey}&skipHttpRedirect=true",
                                 timeout = 10,)

        if not lookupRes.ok:
            print(f"Photo {i} lookup failed ({lookupRes.status_code}): {lookupRes.text}")
            continue
        photoUri = lookupRes.json()["photoUri"]
        mediaRes = requests.get(photoUri, timeout = 10,)
        if not mediaRes.ok:
            print(f"Photo {i} download failed ({mediaRes.status_code}), skipping")
            continue
        contentType = mediaRes.headers.get('Content-Type', 'image/jpeg')
        image_bytes = mediaRes.content
        path = f"{placeId}/{i}.jpg"
        try:
            supabase.storage.from_("place-photos").upload(
                path,
                image_bytes,
                {"contentType": contentType, "upsert": "true"})
        except Exception as e:
            print(f"Photo {i} upload failed: {e}")
            continue
        # Get public URL
        data = supabase.storage.from_("place-photos").get_public_url(path)
        urls.append(data)

    return urls

def map_reviews(reviews: dict) -> dict:
    """
    Map Google Place reviews to a simplified format.

    Args:
        reviews (dict): A list of review dictionaries from the Google Places API.

    Returns:
        list: A list of simplified review dictionaries.
    """
    mapped_reviews = []
    reviews = reviews or []
    for review in reviews[:5]:
        text_obj = review.get("text") or review.get("originalText") or {}
        mapped_reviews.append({
            "author": (review.get("authorAttribution") or {}).get("displayName", "Anonymous"),
            "rating": review.get("rating"),
            "text": text_obj.get("text", ""),
            "relative_time": review.get("relativePublishTimeDescription", "")
        })
    return mapped_reviews

def extract_dishes_and_cuisine(place_name: str, description: str, editorial_summary: str, reviews: list) -> dict:
    """
    Ask Claude for top dishes (from reviews) and a cuisine label (from Google's editorial summary + reviews).

    Returns:
        dict: {"dishes": list[str], "cuisine": str | None}
    """
    review_text = "\n\n".join(
        f"Review {i+1} ({review.get('rating') or '?'}★): {review.get('text')}"
        for i, review in enumerate(reviews)
    )
    context = "\n\n".join(
        part
        for part in [
            f"Editorial description: {description}" if description else "",
            f"Google's summary: {editorial_summary}" if editorial_summary else "",
            f"Google reviews:\n\n{review_text}" if review_text else "",
        ]
        if part
    )
    if not context:
        return {"dishes": [], "cuisine": None}

    prompt = (
        f'Here is information about "{place_name}", a San Francisco restaurant/cafe:\n\n{context}\n\n'
        'Respond with ONLY a JSON object, nothing else, shaped exactly like:\n'
        '{"dishes": ["Pad See Ew", "Thai Iced Tea"], "cuisine": "Thai"}\n'
        '- "dishes": up to 5 specific dishes or menu items mentioned positively in the reviews. Use [] if none are named.\n'
        '- "cuisine": the single best short cuisine label (e.g. "Thai", "Italian", "New American", "Filipino", "Cafe", "Bakery"). Use null if genuinely unclear.'
    )
    headers = {
      "content-type": "application/json",
      "x-api-key": anthropicApiKey,
      "anthropic-version": "2023-06-01",
    }
    data = {
        "model": "claude-sonnet-5",
        "max_tokens": 250,
        "messages": [
            {"role": "user", "content": prompt}
        ]
    }
    res = requests.post("https://api.anthropic.com/v1/messages", headers=headers, json=data, timeout = 30)
    if not res.ok:
        print(f"Dish/cuisine extraction failed ({res.status_code}): {res.text}")
        return {"dishes": [], "cuisine": None}
    result = res.json()
    content = result.get("content") or [{}]
    raw = content[0].get("text") or "{}"
    raw = re.sub(r"^```(?:json)?\s*|\s*```$", "", raw)
    try:
        parsed = json.loads(raw)
        if not isinstance(parsed, dict):
            return {"dishes": [], "cuisine": None}
        dishes = [d for d in parsed.get("dishes", []) if isinstance(d, str)]
        cuisine_val = parsed.get("cuisine")
        cuisine = cuisine_val.strip() if isinstance(cuisine_val, str) and cuisine_val.strip() else None
        return {"dishes": dishes, "cuisine": cuisine}
    except json.JSONDecodeError:
        print(f"  couldn't parse dish/cuisine response: {raw}")
        return {"dishes": [], "cuisine": None}

try:
    response = (supabase.table("places").select("id, name, neighborhood, description, image, cuisine").is_("enriched_at", "null").execute())
       
except Exception as e:
    print(f"Failed to load places: {e}")
    sys.exit(0)

places = response.data

if not places:
    print("No places to enrich.")
    sys.exit(0)

# Enrich each place in the places list
print(f"Enriching {len(places)} places...")

for place in places:
    print(f"- {place['name']}")
    try:
        matched_place = search_google_places(place)
        if not matched_place:
        # Leave enriched_at null — a brand-new place may not be indexed by
        # Google yet, so let the next scan try again instead of skipping it
        # forever.
            print(f"No Google Place found, will retry next run.")
            continue
        if not matched_place.get("photos") or not matched_place.get("reviews"):
            print("  matched but missing photos/reviews, will retry next run")
            continue
        reviews = map_reviews(matched_place.get("reviews"))
        editorial_summary = (matched_place.get("editorialSummary") or {}).get("text")
        photos = upload_photos(place.get("id"), matched_place.get("photos"))
        extracted = extract_dishes_and_cuisine(
            place["name"], place.get("description"), editorial_summary, reviews
        )
        top_dishes = extracted["dishes"]
        # Prefer Google's structured cuisine type; fall back to Claude's read of
        # the description/summary/reviews for the many places Google only tags
        # `restaurant`. Left None if neither is confident.
        cuisine = cuisine_from_google(matched_place) or extracted["cuisine"]

        # a flag indicating whether we got usable data
        got_usable_data = len(photos) > 0 and len(reviews) > 0

        # create json data object for the updated data
        update_data = {
        "google_place_id": matched_place.get("id"),
        "google_maps_uri": matched_place.get("googleMapsUri"),
        "rating" : matched_place.get("rating"),
        "user_rating_count" : matched_place.get("userRatingCount"),
        "reviews" : reviews,
         "photos" : photos,
        "top_dishes" : top_dishes,
        }
    # Only fill the list-card image / cuisine if one wasn't already curated
    # (e.g. by hand in data.json or corrected in the DB) — don't clobber on a re-enrich.
        if not place.get("image"):
            update_data["image"] = photos[0] if photos else None
        if not place.get("cuisine"):
            update_data["cuisine"] = cuisine
        # Update the place in Supabase with the enriched data
        if got_usable_data:
            update_data["enriched_at"] = datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
            print(f"  saved ({len(photos)} photos, {len(reviews)} reviews, {len(top_dishes)} dishes, cuisine: {cuisine or '—'})")
        update_response = (supabase.table("places").update(update_data)
                       .eq("id", place["id"]).execute())
    except Exception as e:
            print(f"  Error enriching place {place['name']}: {e}")
            continue

print("Done.")